Update `@wordpress/components` package's contributing guidelines

ciampo opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
<!-- Please describe what you have changed or added -->

Update the `README` and `CONTRIBUTING` files in the `@wordpress/components` package to reflect the latest guiding principles used by the package maintainers.

Next steps (to be carried out in separate PRs):

- @diegohaz to work on:
  - [ ] Expand composition section
    - [ ] Composition patterns
    - [ ] (Semi-)Controlled components
    - [ ] Layout "responsibilities"
  - [ ] Expand API consistency section
  - [ ] Potentially add a "performance" section (@diegohaz)

- @ciampo to work on:
  - integrate the CONTRIBUTING and README files under `packages/components/src/` into the new guildlines
  - add a style guide on how to write documentation

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

Human reading!

## Screenshots <!-- if applicable -->

N/A

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

New feature (updated / expanded docs)

## Checklist:
- N/A My code is tested.
- N/A My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- N/A My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- N/A I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- N/A My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- N/A I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

ciampo committed:
README: add notes

ciampo committed:
README: add link to `CONTRIBUTING.md`

ciampo committed:
CONTRIBUTING: Add initial notes

ciampo committed:
More notes into CONTRIBUTING

ciampo committed:
Update CONTRIBUTING

ciampo commented:
cc @griffbrad @sarayourfriend @diegohaz 

I've started writing down a draft for an update CONTRIBUTING guidelines for the `@wordpress/components` package.

It'd be great if you could give it a read and help expand on a few sections that, for now, I've only written a few notes for.

sarayourfriend commented on packages/components/CONTRIBUTING.md:
```diff
@@ -4,12 +4,133 @@ Thank you for taking the time to contribute.
+
+[To be expanded] E.g.:
+
+- Boolean component props should be prefixed with `is*` (e.g. `isChecked`) or `has*` (`hasValue`).
```
Should this also include the `enable` prefix?

ciampo replied:
Done in [`e541a7b` (#33960)](https://github.com/WordPress/gutenberg/pull/33960/commits/e541a7b61aaa925921214c8bf80ed8ed2320824e)

ciampo commented on packages/components/CONTRIBUTING.md:
```diff
@@ -4,12 +4,133 @@ Thank you for taking the time to contribute.
+
+UI Primitives can be useful when in need of building an app with all new UI that looks and feels like WordPress but isn't powered by other WordPress technologies (like `@wordpress/i18n`).
+
+_[Note: we need to discuss how to properly classify/label these components. Should we introduce `primitives`, `blocks` and `utils` folders in the `src` folder? Should we include a standard way in each component's README to show which category it belongs to?]_
```
Marking for visibility

ciampo replied:
The potential classification of the components into different groups won't be discussed / applied in this PR. Instead, it can be discussed at a later time in this dedicated issue: #34305

ciampo commented on packages/components/CONTRIBUTING.md:
```diff
@@ -4,12 +4,133 @@ Thank you for taking the time to contribute.
+
+### Components composition
+
+[To be expanded] E.g.:
```
This section needs to be expanded

ciampo commented on packages/components/CONTRIBUTING.md:
```diff
@@ -4,12 +4,133 @@ Thank you for taking the time to contribute.
+
+### APIs Consinstency
+
+[To be expanded] E.g.:
```
This section needs to be expanded

ciampo commented on packages/components/CONTRIBUTING.md:
```diff
@@ -4,12 +4,133 @@ Thank you for taking the time to contribute.
+
+#### Context system
+
+[To be expanded]
```
This section needs to be expanded

ciampo commented on packages/components/CONTRIBUTING.md:
```diff
@@ -4,12 +4,133 @@ Thank you for taking the time to contribute.
+
+#### Hooks vs Components
+
+[To be expanded]
```
This section needs to be expanded

sarayourfriend commented on packages/components/CONTRIBUTING.md:
```diff
@@ -4,12 +4,133 @@ Thank you for taking the time to contribute.
+│   └── index.js
+├── context.ts
+├── index.ts
+├── styles.ts
+└── types.ts
```
It always weirded me out that we split components into separate folders but cram all the types and styles into single files... why is that? Why not split the styles along with the components that use them?

ciampo replied:
That's what I have always thought too — why shouldn't each component have its own `styles.ts` and `types.ts` types?

I have updated the guidelines to follow this pattern in [`6251980` (#33960)](https://github.com/WordPress/gutenberg/pull/33960/commits/6251980121a82b4e8b1a2056ccd0f50afb592649)

sarayourfriend commented on packages/components/CONTRIBUTING.md:
```diff
@@ -4,12 +4,133 @@ Thank you for taking the time to contribute.
+
+- Boolean component props should be prefixed with `is*` (e.g. `isChecked`) or `has*` (`hasValue`).
+- Event callback props should be prefixed with `on*` (e.g. `onChanged`)
+- `styled` components should be suffixed with `*View` (e.g. `BoxView`)
```
This seems like a bad guideline to me now. `const RangeControlView = styled( RangeControl )` doesn't make any sense to me. In what sense is it a "View"?

ciampo replied:
I was under the impression that this pattern had been agreed in previous PRs as a way to recognise `styled` components — here's an [example](https://github.com/WordPress/gutenberg/pull/33054/files#diff-3ba4eeae54a94e8a8f295d9d9052048a33ef16f5f69880ac69dc4dd6138dd752R28).

Anyway, I removed this line in [`aea7c60` (#33960)](https://github.com/WordPress/gutenberg/pull/33960/commits/aea7c603b2ad7f1cca740d667cb1dcab273a84cc)

sarayourfriend replied:
Right, I think we'd discussed it, I'm just rechallenging it after more experience with it before it becomes codified 🙂

sarayourfriend commented on packages/components/CONTRIBUTING.md:
```diff
@@ -4,12 +4,133 @@ Thank you for taking the time to contribute.
+
+#### TypeScript
+
+All new components should be written in TypeScript and should be typed using the `PolymorphicComponent` type (more details about polymorphism can be found above in the "Components composition" section).
```
Should we change the name of PolymorphicComponent soon? WordPressComponent might be a good option? Given that that type can represent non-polymorphic components the name is extremely confusing, especially when used for typing existing components.

ciampo replied:
Agreed. I will work towards renaming that type in a separate PR, then update the guidelines in this PR

gziolo replied:
Until the package is fully typed with TS is rather difficult to enforce this rule. For now, you could word it as "we strongly encourage using TS".

ciampo replied:
Changed the working to "We strongly encourage" and updated from `PolymorphicComponent` to `WordPressComponent` in [`aa5df68` (#33960)](https://github.com/WordPress/gutenberg/pull/33960/commits/aa5df68a656ef6e923107b491bcc8d31feee6c4c)

sarayourfriend commented on packages/components/CONTRIBUTING.md:
```diff
@@ -4,12 +4,133 @@ Thank you for taking the time to contribute.
+
+#### UI Primitives
+
+Lower-level components, which are usually composed together to build the second type of components: UI Blocks. UI Primitives usually don't contain translated text and aim at being highly configurable (at the expense of exposing more props and therefore more complexity).
```
@mtias has different terms for this same concept I believe.

youknowriad replied:
> at the expense of exposing more props and therefore more complexity

This worries me a bit, we could quickly regret it 🤷‍♂️ I think if we can keep these "primitives" internal, it's better (not expose them in wp.components).

Also, we already have a concept of "primitives": components that replace DOM elements to be mobile compatible `@wordpress/primitives`

ciampo replied:
> @mtias has different terms for this same concept I believe.
> Also, we already have a concept of "primitives"

I'm completely fine with using different terminology (rather than `Primitives` and `UI blocks`). The most important thing to me is that we (if we want to) introduce a distinction between low-level and high-level components, which may help users of this package (including other Gutenberg developers) to choose the right component for a given task.

Does anyone have any suggestions on the naming?

> This worries me a bit, we could quickly regret it 🤷‍♂️

I'm fine with removing the `at the expense of exposing more props and therefore more complexity` text — the main reason why I added it is because it had been previously suggested as a potential difference between "low" and "high" level components.

> I think if we can keep these "primitives" internal, it's better (not expose them in wp.components).

I think that low-level components can still be useful — for example, in the case a user wanted to create their own version of a UI component, but without re-implementing the functionality from scratch and while keeping the WordPress look&feel.

ciampo replied:
This topic can be discussed at a later time in this dedicated issue: #34305

In the meantime, I will go ahead and remove any text referencing any potential components classification in this PR

sarayourfriend commented on packages/components/CONTRIBUTING.md:
```diff
@@ -4,12 +4,133 @@ Thank you for taking the time to contribute.
+
+_[Note: we need to discuss how to properly classify/label these components. Should we introduce `primitives`, `blocks` and `utils` folders in the `src` folder? Should we include a standard way in each component's README to show which category it belongs to?]_
+
+#### UI Blocks
```
Blocks is a very bad name to use, in my opinion. In the context of WordPress and Gutenberg especially it already means something very specific which the components package has nothing to do with.

ciampo replied:
I've chosen Primitives/Blocks for the first draft version of the guidelines based on this open issue (#33111), but I agree that we should pick a better (less confusing) name.

ciampo replied:
The conversation can continue in https://github.com/WordPress/gutenberg/pull/33960#discussion_r695054513.

sarayourfriend commented on packages/components/CONTRIBUTING.md:
```diff
@@ -4,12 +4,133 @@ Thank you for taking the time to contribute.
+
+The `@wordpress/components` package includes several components which are relied upon by many developers across different projects. It is, therefore, very important to avoid introducing breaking changes as much as possible.
+
+In these situations, an alternative approach is to "soft-deprecate" the given legacy API. This is achieved by removing traces of the API from the docs, while the code keeps supporing it in the background (with or without a warning)
```
When and why do we add or not add a deprecation warning?

ciampo replied:
That's a great question and one that I don't have an answer to. That's why it would be great to collaborate and come up with a more objective way to determine what do to in these scenarios.

ciampo replied:
cc @gziolo in case you have highlights about the soft deprecation strategy, in particular about if and when we should output a deprecation warning

gziolo replied:
We need to update first all places in Gutenberg that use API that is about to be deprecated. When that is ready, we can add warnings. If we do it earlier then Browser Console will be printing all those warnings. In addition to that, it would also fail e2e tests.

ciampo replied:
Thank you for the explanation! I've amended the paragraph and added those "steps" in [`37c949d` (#33960)](https://github.com/WordPress/gutenberg/pull/33960/commits/37c949dfe3421109c6d8663e67dd2cca60e292dc)

ciampo committed:
Add `enable` to boolean prefixes

ciampo committed:
Remove `View` suffix for styled components

ciampo committed:
Update folder structure: move styles and types into each subfolder

gziolo commented on packages/components/CONTRIBUTING.md:
```diff
@@ -4,12 +4,134 @@ Thank you for taking the time to contribute.
--   It has to define a React component called `My<ComponentName>` which renders the example (i.e.: `MyButton`). Examples for the Higher Order Components should define a `MyComponent<ComponentName>` component (i.e.: `MyComponentWithNotices`).
+### Compatibility
+
+The `@wordpress/components` package includes several components which are relied upon by many developers across different projects. It is, therefore, very important to avoid introducing breaking changes as much as possible.
```
I don't think we want to have any breaking changes for stable APIs:

```suggestion
The `@wordpress/components` package includes components that are relied upon by many developers across different projects. It is, therefore, very important to avoid introducing breaking changes.
```

The suggested alternative to soft-deprecate API is the way to go.

ciampo replied:
I know that this PR is not the right place for this conversation to happen, but in the future I'd like to discuss how we can prepare a strategy for introducing breaking changes to this package.

For now, we can definitely go with what you're advocating (thank you for the suggestion!)

youknowriad commented on packages/components/CONTRIBUTING.md:
```diff
@@ -4,12 +4,134 @@ Thank you for taking the time to contribute.
+
+#### Unit tests
+
+All new components should include unit tests using [testing library](https://testing-library.com/). Ideally, the component's main piece of functionality would be tested explicitely (rather than using snapshots).
```
Is it really necessary to have mandatory unit tests for components. Usually they turned out to be testing implementation details.

ciampo replied:
> Usually they turned out to be testing implementation details.

I agree that most unit tests are either snapshots or usually poorly written. Ideally, it should be the reviewer's duty to ensure that the unit tests added to the codebase are well written.

Is not requesting unit tests a better solution?

ciampo replied:
Update: I merged #34423, rebased this PR, and updated the "Unit Tests" section in 0a36f58 (#33960) as agreed in [this conversation](https://github.com/WordPress/gutenberg/pull/33960#discussion_r690196895).

There's no mention of "required" unit tests anymore, but I'd be interested to know your opinion on my answer above.

gziolo commented on packages/components/CONTRIBUTING.md:
```diff
@@ -4,12 +4,134 @@ Thank you for taking the time to contribute.
+
+### Components structure
+
+The contents of the `@wordpress/components` package can be divided into three groups: UI Primitives, UI Blocks, and Utilities.
```
The proposed structure isn't reflected in the code as of today (as noted below). It is confusing when you don't have a deep understanding of the subtle differences between UI primitives or UI blocks. At the end of the day, what is the most important is what gets exposed as public API. I don't think we ever want to expose utilities as they should rather be used to help with code reuse internally.

ciampo replied:
I replied to a similar point made by @priethor in [this other comment](https://github.com/WordPress/gutenberg/pull/33960#discussion_r695093491). Let's move this conversation there

gziolo commented on packages/components/CONTRIBUTING.md:
```diff
@@ -4,12 +4,134 @@ Thank you for taking the time to contribute.
+
+All new component should be styled using [Emotion](https://emotion.sh/docs/introduction).
+
+Note: Instead of using Emotion's standard `cx` function, the custom [`useCx` hook](/packages/components/src/utils/hooks/use-cx.ts) should be used instead.
```
Can we add an ESLint rule that helps to catch those violations?

sarayourfriend replied:
There already is, importing from `@emotion/css` where `cx` lives is a restricted import. @ciampo we might need to update the message that accompanies that rule though.

ciampo replied:
I'll look into the ESLint rule message and report updates here

ciampo replied:
I've opened #34418 to update the message — feel free to leave any feedback there

gziolo commented on packages/components/CONTRIBUTING.md:
```diff
@@ -4,12 +4,134 @@ Thank you for taking the time to contribute.
+
+#### Unit tests
+
+All new components should include unit tests using [testing library](https://testing-library.com/). Ideally, the component's main piece of functionality would be tested explicitely (rather than using snapshots).
```
It should be expanded with some examples. Otherwise, it isn't very valuable for folks who don't know what testing library is or what snapshot testing is. We can reference more detailed documentation:

https://github.com/WordPress/gutenberg/blob/trunk/docs/contributors/code/testing-overview.md#javascript-testing

There is also a section about using snapshots. The challenge is that it still references Enzyme so it should get updated with the recommendations you have here.

ciampo replied:
If I understand correctly, do you suggest that we:

1. Update the Testing Overview docs to reference Testing Library (should it be in addition to the Enzyme example, or instead of the Enzyme example?)
2. In this PR, reference to the Testing Guidelines

If that's the case, I can look into updating the Testing Overview docs in a separate PR.

gziolo replied:
Let's pretend we don't use Enzyme at all 😅 

I think you sketched a good plan, in this PR we can reference testing guidelines and tackle up to date testing recommendation separately 👍🏻

ciampo replied:
Opened #34423 for the changes to the Testing Overview docs

ciampo replied:
merged #34423, rebased this PR, and updated the "Unit Tests" section in [`0a36f58` (#33960)](https://github.com/WordPress/gutenberg/pull/33960/commits/0a36f58417cb3f10434d6a040a7b2c9d3520d0ec)

gziolo commented on packages/components/CONTRIBUTING.md:
```diff
@@ -4,12 +4,134 @@ Thank you for taking the time to contribute.
+
+#### Storybook
+
+All new components should include Storybook examples.
```
It needs more details because more contributors won't know how to do it.

ciampo replied:
Makes sense! I will add more details to this section

ciampo replied:
Added more details in [`4458f85` (#33960)](https://github.com/WordPress/gutenberg/pull/33960/commits/4458f8502ce312c89bc61c3186fa2a4bdf5c36b3)

gziolo reviewed (commented):
That's a great initiative. Thank you for starting looking at bringing some structure and better organization to `@wordpress/components` 👏🏻

priethor commented on packages/components/README.md:
````diff
@@ -10,6 +10,7 @@ Install the module
 npm install @wordpress/components --save
 ```
 
+[Note: is the following paragraph still relevant?]
````
We can probably remove this as IE is not supported in the editor anymore.

gziolo replied:
We polyfill some functionalities in WP core even after we dropped support for IE 11. We need to reword it and link to how to use the polyfill shipped in `@wordpress/babel-preset-default`:

https://github.com/WordPress/gutenberg/tree/trunk/packages/babel-preset-default#polyfill

It's more about the most recent features that are part of the JavaScript language but some browser vendors still didn't implement them. See in WP core:

https://github.com/WordPress/WordPress/blob/master/wp-includes/js/dist/vendor/wp-polyfill.js

ciampo replied:
I will look into this in a separate PR and report back here with updates.

ciampo replied:
I will look into this in a separate PR and report back here with updates.

ciampo replied:
Didn't need a separate PR. Updated the "polyfills" paragraph with references to the polyfill shipped in `@wordpress/babel-preset-default` in [`4409c6c` (#33960)](https://github.com/WordPress/gutenberg/pull/33960/commits/4409c6c9b394c04ae6b4774860212eaa7bfdec1e)

priethor commented on packages/components/CONTRIBUTING.md:
```diff
@@ -4,12 +4,134 @@ Thank you for taking the time to contribute.
+
+#### Folder structure
+
+Following all previous guidelines, all new components should follow this folder structure:
```
Does this structure apply to UI-ess utilities as well? As @gziolo mentioned above, it doesn't seem like these utilities need to be exposed as they are reusable implementation details rather than components to be consumed externally. If so, I'd lean towards separating utilities in the folder structure.

ciampo replied:
> it doesn't seem like these utilities need to be exposed as they are reusable implementation details rather than components to be consumed externally

That's a good point! I guess it greatly depends on what we define as a "utility" — see for example this [draft list](https://github.com/WordPress/gutenberg/issues/33111#issuecomment-871569331) that @sarayourfriend and I worked on some time ago.

If we were to go with this list, I would argue that some of these "utilities" could be useful if exported.

> I'd lean towards separating utilities in the folder structure.

If you refer to a component's folder structure, then I agree that we could change the wording so that the suggested folder structure _does not_ apply to internal utility components/function.

With regards to the overall package folder structure, I left a [note](https://github.com/WordPress/gutenberg/pull/33960#discussion_r687024506) in the `CONTRIBUTING.md` file in this PR about it — would love to hear your opinion about it too!

ciampo replied:
Hey @priethor , just a quick update — I've created a separate issue (#34305) to discuss the potential classification of the components into different groups.

I've also slightly reworded the first paragraph of the "Folder Structure" section in [`5111e1e` (#33960)](https://github.com/WordPress/gutenberg/pull/33960/commits/5111e1e17a671b7cb34db7fe148adaced27fd4d0) — it now specifies that this folder structure should not apply to shared utilities (i.e. not components).

ciampo committed:
Less strict working about typescript, rename PolymorphicComponent to WordPressComponent

ciampo committed:
Update packages/components/CONTRIBUTING.md

Co-authored-by: Greg Ziółkowski <grzegorz@gziolo.pl>

ciampo committed:
Refer to the repo testing overview docs in the unit test section

ciampo committed:
Update polyfill info

ciampo committed:
Remove "Components Structure" paragraph

ciampo committed:
Reword text in "Folder structure" section, exclude  shared utilities

ciampo committed:
Folder structure schemes: name parent folders, remove useless sub-component folder

ciampo committed:
Expand Storybook section

ciampo committed:
Add details to soft deprecation strategy

ciampo committed:
Misc smaller improvements

ciampo committed:
Fix typo

gziolo commented on packages/components/CONTRIBUTING.md:
```diff
@@ -14,7 +14,11 @@ This set of guidelines should apply especially to newly introduced components. I
 The `@wordpress/components` package includes components that are relied upon by many developers across different projects. It is, therefore, very important to avoid introducing breaking changes.
 
-In these situations, an alternative approach is to "soft-deprecate" the given legacy API. This is achieved by removing traces of the API from the docs, while the code keeps supporing it in the background (with or without a warning)
+In these situations, one possible approach is to "soft-deprecate" a given legacy API. This is achieved by:
```
Looks correct 👍🏻

ciampo committed:
Expand hooks vs components section, move under composition

ciampo committed:
Move sub components naming convention under "APIs consistency" section

ciampo committed:
Add Polymorphic components sub-section, reorganise "Components composition" section into sub-sections

ciampo committed:
Add more details about listing props through knobs in storybook

ciampo committed:
Remove hyphen from subcomponents

ciampo committed:
Add content to the `Context System` section

gziolo commented:
You might want to split this PR. It's getting big already 😄 Let us know when it's ready for the final review.

ciampo commented:
> Let us know when it's ready for the final review.

It's not quite there yet — yesterday we agreed that @diegohaz is going to contribute to the "composition", "API consistency" and potentially a "performance" section (I've updated the PR description to keep track of these tasks)

> You might want to split this PR. It's getting big already 😄 

That makes sense! Let's wait for the bulk of the work to be done, and then we can make a decision on if/how to split this PR

gziolo commented:
> That makes sense! Let's wait for the bulk of the work to be done, and then we can make a decision on if/how to split this PR

Actually, I was thinking more of merging sooner some parts that you consider ready and continue working on the rest in this PR 😄

ciampo committed:
comment out unfinied sections

ciampo commented:
> > That makes sense! Let's wait for the bulk of the work to be done, and then we can make a decision on if/how to split this PR
> 
> Actually, I was thinking more of merging sooner some parts that you consider ready and continue working on the rest in this PR 😄

Oh, I see! In that case, you may start another round of review. I've just commented out the sections that were not completed ([`7c78840` (#33960)](https://github.com/WordPress/gutenberg/pull/33960/commits/7c78840608430e7ef31848d97879f0d6c56efac3)) — we will work on them in separate PRs:

- @diegohaz to work on:
  - [ ] Expand composition section
    - [ ] Composition patterns
    - [ ] (Semi-)Controlled components
    - [ ] Layout "responsibilities"
  - [ ] Expand API consistency section
  - [ ] Potentially add a "performance" section (@diegohaz)

- @ciampo to work review, potentially integrate and delete the CONTRIBUTING and README files under `packages/components/src/`

diegohaz commented on packages/components/CONTRIBUTING.md:
```diff
@@ -4,12 +4,202 @@ Thank you for taking the time to contribute.
+Contributions to the `@wordpress/components` package should follow a set of core principles and technical requirements.
 
-These examples can be consumed automatically from other projects in order to visualize them in their documentation. To ensure these examples are extractable, compilable and renderable, they should be structured in the following way:
+This set of guidelines should apply especially to newly introduced components. It is, in fact, possible that some of the older component don't respect some of these guidelines for legacy/compat reasons.
```
```suggestion
This set of guidelines should apply especially to newly introduced components. It is, in fact, possible that some of the older components don't respect some of these guidelines for legacy/compatibility reasons.
```

diegohaz commented on packages/components/CONTRIBUTING.md:
```diff
@@ -4,12 +4,202 @@ Thank you for taking the time to contribute.
+
+1. Removing traces of the API from the docs, while still supporing it in code;
+2. Updating all places in Gutenberg that use that API;
+3. Adding deprecation warnings (only after the previous point is completed, otherwide the Browser Console will be polluted by all those warnings and some e2e tests may fail).
```
```suggestion
3. Adding deprecation warnings (only after the previous point is completed, otherwise the Browser Console will be polluted by all those warnings and some e2e tests may fail).
```

diegohaz commented on packages/components/CONTRIBUTING.md:
```diff
@@ -4,12 +4,202 @@ Thank you for taking the time to contribute.
+1. Removing traces of the API from the docs, while still supporing it in code;
+2. Updating all places in Gutenberg that use that API;
+3. Adding deprecation warnings (only after the previous point is completed, otherwide the Browser Console will be polluted by all those warnings and some e2e tests may fail).
+
```
I'd suggest adding a note about prefixing new components and props with `__unstable` or `__experimental`. I'm not sure if we all agree with that, but that's something I often suggest in code reviews.

Also, linking to the specific docs would be useful.

```suggestion

When adding new components or new props to existing components, it's recommended to prefix them with `__unstable` or `__experimental` until they're stable enough to be exposed as part of the public API.

Learn more on [How to preserve backward compatibility for a React Component](https://github.com/WordPress/gutenberg/blob/trunk/docs/how-to-guides/backward-compatibility/README.md#how-to-preserve-backward-compatibility-for-a-react-component) and [Experimental and Unstable APIs](https://github.com/WordPress/gutenberg/blob/trunk/docs/contributors/code/coding-guidelines.md#experimental-and-unstable-apis).

```

gziolo replied:
That's a good default to go with. In the context of components, it's going to be experimental in most of the cases ❤️

ciampo replied:
Makes a lot of sense, thank you @diegohaz for suggesting this approach!

@gziolo, out of curiosity — when adding links to other files in the repo, I omitted the `(https://github.com/WordPress/gutenberg/blob/trunk/` prefix.

Do you have a preference around the link format?

gziolo replied:
At some point @sarayourfriend updated all links to start with: `https://github.com/WordPress/gutenberg/blob/HEAD/`. It might be related to the switch from `master` to `trunk` though.

Some links get special treatment in the Block Editor Handbook, see more in https://github.com/WordPress/gutenberg/tree/trunk/docs/contributors/documentation#using-links.

sarayourfriend replied:
Yeah it was related to the branch rename... HEAD is a neutral pointer to the "main" branch of a repository regardless of the name, so we were able to complete that before the tools for renaming the branch were available (basically pre-work we could do to prepare the repo for the rename).

We can use the `trunk` name directly I guess, but if the branch name ever changed again for some reason you'd incur the cost of doing the update to HEAD again.

ciampo replied:
Thank you both for the details explanation!

Following the documentation linked by @gziolo , I've decided to update all internal link to the "absolute path" format in [`8c494c8` (#33960)](https://github.com/WordPress/gutenberg/pull/33960/commits/8c494c83282cb1a9cf68f0bb51597d91d5f1abb5)

diegohaz commented on packages/components/CONTRIBUTING.md:
```diff
@@ -4,12 +4,202 @@ Thank you for taking the time to contribute.
+// in `hook.ts`
+function useExampleComponent( props: PolymorphicComponentProps< ExampleProps, 'div' > ) {
+	// Merge received props with the context system.
+	const { isVisible, className, ...otherProps	} = useContextSystem( props, 'Example' );
```
```suggestion
	const { isVisible, className, ...otherProps } = useContextSystem( props, 'Example' );
```

gziolo replied:
Is Prettier ignoring `tsx` in MD files?

ciampo replied:
I don't personally remember Prettier ever auto-formatting code snippets in MD files — has that always been supposed to happen?

sarayourfriend replied:
Prettier is definitely able to do it: https://prettier.io/blog/2017/11/07/1.8.0.html

> Powered by Prettier's generic "multiparser", Prettier will format code blocks in Markdown! We use the language code provided with the code block to determine which language it is, and thus we can format any language that Prettier supports (including Markdown itself, if you're into that).

It might be slipping up on the `tsx` definition? Or could just be a configuration issue, or if Prettier isn't enabled on MD files for some reason.

ciampo replied:
It doesn't seem to work for me even for `jsx` snippets — does it work for you?

diegohaz commented on packages/components/CONTRIBUTING.md:
```diff
@@ -4,12 +4,202 @@ Thank you for taking the time to contribute.
+
+The following are a set of technical requirements for all newly introduced components. These requirements are also retroactively being applied to existing components.
+
+For an example of a component that follows these requirements, take a look at [`ItemGroup`](/packages/components/src/item-group) as an example.
```
```suggestion
For an example of a component that follows these requirements, take a look at [`ItemGroup`](/packages/components/src/item-group).
```

diegohaz commented on packages/components/CONTRIBUTING.md:
```diff
@@ -4,12 +4,202 @@ Thank you for taking the time to contribute.
+
+For even more flexibility while writing stories, the [`@storybook/addons-controls`](https://storybook.js.org/addons/@storybook/addon-controls) addon is a great way to interact with a component's arguments dynamically and in real time.
+
+Ideally, all of the props for the story's components should be exposed by using knobs. The default value of the knobs should coincide with the default value of the props (i.e. it should be `undefined` if a prop is not required). A story should, therefore, also explicitely show how values from the Context System are applied to (sub)components. A good example of how this may look like is the [`Card ...
```
```suggestion
Ideally, all of the props for the story's components should be exposed by using knobs. The default value of the knobs should coincide with the default value of the props (i.e. it should be `undefined` if a prop is not required). A story should, therefore, also explicitly show how values from the Context System are applied to (sub)components. A good example of how this may look like is the [`Card` story](https://wordpress.github.io/gutenberg/?path=/story/components-card--default) (code [here](/packages/components/src/card/stories/index.js)).
```

diegohaz reviewed (commented):
Great work, @ciampo! Thank you for working on this. I made a few minor suggestions and corrections.

gziolo commented on packages/components/CONTRIBUTING.md:
```diff
@@ -4,12 +4,202 @@ Thank you for taking the time to contribute.
+In these situations, one possible approach is to "soft-deprecate" a given legacy API. This is achieved by:
+
+1. Removing traces of the API from the docs, while still supporing it in code;
+2. Updating all places in Gutenberg that use that API;
```
```suggestion
1. Removing traces of the API from the docs, while still supporting it in code.
2. Updating all places in Gutenberg that use that API.
```

gziolo commented on packages/components/CONTRIBUTING.md:
```diff
@@ -4,12 +4,202 @@ Thank you for taking the time to contribute.
+
+#### TypeScript
+
+We strongly encourage using TypeScript for all new components. Components should be typed using the `WordPressComponent` type (more details about polymorphism can be found above in the "Components composition" section).
```
There is no "Components composition" section yet so the part in parenthesis should be commented out for now.

ciampo replied:
Done in [`6647446` (#33960)](https://github.com/WordPress/gutenberg/pull/33960/commits/66474469b39466210d2596af1110719ea4e84602)

gziolo commented on packages/components/CONTRIBUTING.md:
```diff
@@ -4,12 +4,202 @@ Thank you for taking the time to contribute.
+
+We strongly encourage using TypeScript for all new components. Components should be typed using the `WordPressComponent` type (more details about polymorphism can be found above in the "Components composition" section).
+
+#### Emotion
```
Should this section be renamed to "Styling"? If someone doesn't know Emotion, they might miss this section when the only skim the document.

ciampo replied:
Good point. Done in [`670ca6c` (#33960)](https://github.com/WordPress/gutenberg/pull/33960/commits/670ca6c4bccbd151dc87908f95669152dc257fcd)

gziolo commented on packages/components/CONTRIBUTING.md:
```diff
@@ -4,12 +4,202 @@ Thank you for taking the time to contribute.
+- they can connect to the Context via `contextConnect`
+- they can read the "computed" values from the context via `useContextSystem`
+
+An example of how this is used is the [`Card` component](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/card/card/component.js#L44-L54) — all of its subcomponents (e.g. `CardBody`, `CardHeader`...) will [use, by default, the same value for the `size` as the parent `Card` component](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/card/card-body/ ...
```
The link to the lines `Card` and `CardBody` components might quickly get out misleading when lines are added or removed. Can we copy this snippet here and link to the files instead?

ciampo replied:
I've updated this section by adding a code snippet showing how the `CardBody` component "inherits" the `size` prop from the parent `Card` via the Context System in [`c34f731` (#33960)](https://github.com/WordPress/gutenberg/pull/33960/commits/c34f73168aafb25ab9412d3f0c4f5ed59c3c1358)

gziolo commented on packages/components/CONTRIBUTING.md:
```diff
@@ -4,12 +4,202 @@ Thank you for taking the time to contribute.
+
+#### Unit tests
+
+Please refer to the [JavaScript Testing Overview docs](/docs/contributors/code/testing-overview.md).
```
It would be better to link to the section about testing components.

ciampo replied:
I wasn't sure about exactly what section you were referring to — for now I've updated the link to point to the snapshot testing in [`e65518d` (#33960)](https://github.com/WordPress/gutenberg/pull/33960/commits/e65518d1c9553d02ea31186602eb3fe18206863c)

gziolo commented on packages/components/CONTRIBUTING.md:
```diff
@@ -4,12 +4,202 @@ Thank you for taking the time to contribute.
+
+#### Storybook
+
+All new components should include [Storybook](https://storybook.js.org/) stories, in order to make it easier to work on and review each component in isolation.
```
Can we expand on what the story is?

ciampo replied:
Done in [`b6edc2e` (#33960)](https://github.com/WordPress/gutenberg/pull/33960/commits/b6edc2e0ddcd88ca1210bd8565b34a16d874791d)

gziolo commented on packages/components/CONTRIBUTING.md:
```diff
@@ -4,12 +4,202 @@ Thank you for taking the time to contribute.
+
+For even more flexibility while writing stories, the [`@storybook/addons-controls`](https://storybook.js.org/addons/@storybook/addon-controls) addon is a great way to interact with a component's arguments dynamically and in real time.
+
+Ideally, all of the props for the story's components should be exposed by using knobs. The default value of the knobs should coincide with the default value of the props (i.e. it should be `undefined` if a prop is not required). A story should, therefore, also explicitely show how values from the Context System are applied to (sub)components. A good example of how this may look like is the [`Card ...
```
Can we explain also what the "knob" is?

ciampo replied:
Done in [`b6edc2e` (#33960)](https://github.com/WordPress/gutenberg/pull/33960/commits/b6edc2e0ddcd88ca1210bd8565b34a16d874791d) — I stopped using the term "knob", and instead I'm referring to the Storybook Controls addon (which replaced knobs). I've also added some more details on what controls are and what they allow

gziolo commented on packages/components/CONTRIBUTING.md:
```diff
@@ -4,12 +4,202 @@ Thank you for taking the time to contribute.
+
+#### Documentation
+
+All components, in addition to being typed, should be using JSDoc when necessary.
```
We should link to coding guidelines and JSDoc section:

https://github.com/WordPress/gutenberg/blob/trunk/docs/contributors/code/coding-guidelines.md#javascript-documentation-using-jsdoc

By the way, it would be good to revisit if that section didn't get out of date now we use TypeScript so widely.

ciampo replied:
Linked to the docs in [`e8bb18d` (#33960)](https://github.com/WordPress/gutenberg/pull/33960/commits/e8bb18d572c23644a0d35173fa45abaa8038211b).

> it would be good to revisit if that section didn't get out of date now we use TypeScript so widely.

I've taken note and I'll try to review and potentially revisit that section in a separate PR

gziolo commented on packages/components/README.md:
````diff
@@ -27,8 +27,13 @@ export default function MyButton() {
 }
 ```
 
+[Note: is the following paragraph still relevant?]
````
As long as some components still use SCSS files, we need it. In the long run, I assume Emotion would handle all styles and it might become obsolete.

ciampo replied:
Deleted this comment in [`08cd356` (#33960)](https://github.com/WordPress/gutenberg/pull/33960/commits/08cd3568541b4a63fc814d9c99b2c47e1c89ed32)

gziolo commented on packages/components/CONTRIBUTING.md:
````diff
@@ -4,12 +4,202 @@ Thank you for taking the time to contribute.
+└── types.ts
+```
+
+In case of a family of components (e.g. `Card` and `CardBody`, `CardFooter`, `CardHeader` ...), each component's implementation should live in a separate subfolder:
````
It looks complex to apply but I see how it could benefit code readability in the long run. Besides, I like when code is structured the same way creating a pattern you can learn quickly 💯

ciampo replied:
I agree. Having a well defined way to structure components can definitely help to create a feeling of familiarity when working within the components package!

gziolo reviewed (commented):
I also included a few nitpicks. I found some typos and marked a few places where we could link to other places. There are some concepts that we should explain better for folks who don't know the stack that we use. 

Great work so far. This document is very helpful. Let's merge when comments are addressed 👍🏻

ciampo committed:
Apply suggestions from code review

Co-authored-by: Haz <hazdiego@gmail.com>
Co-authored-by: Greg Ziółkowski <grzegorz@gziolo.pl>

ciampo committed:
Comment out reference to the "composition section"

ciampo committed:
Rename "Emotion" section to "Styling"

ciampo committed:
Updale unit test section to link directly to components snapshot testing

ciampo committed:
Update the storybook section, adding more details on stories and knobs

ciampo committed:
Link to the Coding Guidelines in the "Documentation" section

ciampo committed:
remove note about relevance of SCSS-related docs in README

ciampo committed:
Update context section with example snippets

ciampo committed:
Format internal links to absolute path format

ciampo commented:
Thank you @diegohaz and @gziolo for your last round of review! I've rebased the PR and addressed and/or replied to all of your points — I believe that we're going to be able to merge this PR very soon!

gziolo commented on packages/components/README.md:
````diff
@@ -10,7 +10,7 @@ Install the module
 ```
 
-_This package assumes that your code will run in an **ES2015+** environment. If you're using an environment that has limited or no support for ES2015+ such as IE browsers then using [core-js](https://github.com/zloirock/core-js) will add polyfills for these methods._
+_This package assumes that your code will run in an **ES2015+** environment. If you're using an environment that has limited or no support for such language features and APIs, you should include [the polyfill shipped in `@wordpress/babel-preset-default`](/packages/babel-preset-default#polyfill) in your code._
````
Good job with this note’s update. I need to copy it to all other packages 😃

gziolo replied:
Opened https://github.com/WordPress/gutenberg/pull/34878 to propagate the same change to other packages.

gziolo reviewed (approved):
This is looking great. Let’s land and iterate as discussed above.

Excellent work documenting every detail ❤️

diegohaz reviewed (approved):
Awesome! Great work @ciampo!
